### PR TITLE
Lobby

### PR DESCRIPTION
**Pull request by BobvanVelzen**

### CODE REVIEW

File: app/dal/contexts/JPAUserContext.java [hunk 1]
```diff
@@ -11,9 +11,7 @@
     private final JPAApi jpaApi;
 
     @javax.inject.Inject
-    public JPAUserContext(JPAApi jpaApi) {
-        this.jpaApi = jpaApi;
-    }
+    public JPAUserContext(JPAApi jpaApi) { this.jpaApi = jpaApi; }
```
Comment (REXOTIUM) [inline app/dal/contexts/JPAUserContext.java]: No oneliners for this please

File: app/dal/contexts/JPAGameContext.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        List<Game> results = getEntityManager()
+                .createNamedQuery("Game.findFirstOpen", Game.class)
+                .setMaxResults(1)
+                .getResultList();
+        if (results.isEmpty()) return null;
+        else return results.get(0);
+    }
+
+    @Override
+    public void saveGame(Game game) { getEntityManager().persist(game); }
```
Comment (REXOTIUM) [inline app/dal/contexts/JPAGameContext.java]: No oneliners please

File: app/dal/entities/Game.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package dal.entities;
+import javax.persistence.*;
+import java.util.ArrayList;
+import java.util.List;
+
+@Entity
+@Table(name = "Game")
+@NamedQueries({
+        @NamedQuery(name = "Game.getAll", query = "select g from Game as g"),
+        @NamedQuery(name = "Game.findOne", query = "select g from Game as g where id = :id"),
```
Comment (REXOTIUM) [inline app/dal/entities/Game.java]: g.id instead of id

File: app/dal/entities/Game.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    public int getId() {
+        return id;
+    }
+
+    public void setId(int id) {
+        this.id = id;
+    }
+
+    public List<Session> getSessionList() { return this.sessions; }
```
Comment (REXOTIUM) [inline app/dal/entities/Game.java]: No oneliners please

File: app/dal/entities/Game.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    public void addSession(Session session) { this.sessions.add(session); }
```
Comment (REXOTIUM) [inline app/dal/entities/Game.java]: No oneliners please

File: app/dal/entities/Game.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    public void removeSession(Session session) { this.sessions.remove(session); }
```
Comment (REXOTIUM) [inline app/dal/entities/Game.java]: No oneliners please

File: app/dal/repositories/GameRepository.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+
+@Singleton
+public class GameRepository implements GameContext{
+
+    private GameContext context;
+
+    @Inject
+    public GameRepository(GameContext context) {
+
```
Comment (REXOTIUM) [inline app/dal/repositories/GameRepository.java]: Remove useless line return

File: app/dal/repositories/GameRepository.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        this.context = context;
+    }
+
+    @Override
+    public List<Game> getGames() {
+        return context.getGames();
+    }
+
+    @Override
+    public Game getGame(int gameId) { return context.getGame(gameId); }
```
Comment (REXOTIUM) [inline app/dal/repositories/GameRepository.java]: No oneliners please

File: app/dal/repositories/GameRepository.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    public Game getOpenGame() { return context.getOpenGame(); }
```
Comment (REXOTIUM) [inline app/dal/repositories/GameRepository.java]: No oneliners please

File: app/dal/repositories/GameRepository.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    public void saveGame(Game game) { context.saveGame(game); }
```
Comment (REXOTIUM) [inline app/dal/repositories/GameRepository.java]: No oneliners please

File: app/dal/repositories/GameRepository.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    public void saveSession(Session session) { context.saveSession(session); }
```
Comment (REXOTIUM) [inline app/dal/repositories/GameRepository.java]: No oneliners please

**Comment by REXOTIUM** [comment]
Also resolve current conflicts

File: app/controllers/LobbyController.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        Session player = new Session();
+        player.setSessionId(uuid);
+        gameRepository.saveSession(player);
+        game.addSession(player);
+
+        ObjectNode json = mapper.createObjectNode();
+
+        json.put("sessionid", player.getSessionId());
+        json.put("gameid", game.getId());
+        json.put("size", game.getSessionList().size());
```
Comment (C-Alexander) [inline app/controllers/LobbyController.java]: getSessions is probably better naming, don't have to put "list" into the name.

File: app/controllers/LobbyController.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Transactional
+    public Result addToQueue() {
+        JsonNode body = request().body().asJson();
+
+        String uuid = body.get("sessionid").asText();
+
+        Game game = gameRepository.getOpenGame();
+        if (game == null) { game = new Game(); }
+        gameRepository.saveGame(game);
```
Comment (C-Alexander) [inline app/controllers/LobbyController.java]: A getOrCreate function would probably make more sense,  {} isn't needed for one liner if statements (but not wrong either)

File: app/controllers/LobbyController.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+        // Voorkom dubbele sessions
+        Session player = new Session();
```
Comment (C-Alexander) [inline app/controllers/LobbyController.java]: also why name them player

File: app/controllers/LobbyController.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    public LobbyController(GameRepository gameRepository) {
+        this.gameRepository = gameRepository;
+        mapper = new ObjectMapper();
+    }
+
+    @Transactional
+    public Result addToQueue() {
+        JsonNode body = request().body().asJson();
+
+        String uuid = body.get("sessionid").asText();
```
Comment (C-Alexander) [inline app/controllers/LobbyController.java]: Please use the same caps standard as used elsewhere (camelcase), also please dont call it session anything, we already have a class named sessions.

File: app/controllers/LobbyController.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+        Game game = gameRepository.getOpenGame();
+        if (game == null) { game = new Game(); }
+        gameRepository.saveGame(game);
+
+        // Voorkom dubbele sessions
+        Session player = new Session();
+        player.setSessionId(uuid);
+        gameRepository.saveSession(player);
```
Comment (C-Alexander) [inline app/controllers/LobbyController.java]: why save the session in the gameRepository? There's a sessionrepository...

File: app/dal/contexts/JPAGameContext.java (hunk 2 above)
Comment (C-Alexander) [reply app/dal/contexts/JPAGameContext.java]: had ik in de klas ook al gezegd maar ben het er niet mee eens dat dit soort oneliners slecht zijn, gezien intellij ze automatisch inklapt enzo zou het mega irritant zijn om ermee op te houden. 

Wissel nou eens naar een echte ide :P

File: app/dal/entities/Game.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import javax.persistence.*;
+import java.util.ArrayList;
+import java.util.List;
+
+@Entity
+@Table(name = "Game")
+@NamedQueries({
+        @NamedQuery(name = "Game.getAll", query = "select g from Game as g"),
+        @NamedQuery(name = "Game.findOne", query = "select g from Game as g where id = :id"),
+        @NamedQuery(name = "Game.findFirstOpen", query = "select g from Game as g where (SELECT COUNT(*) FROM g.sessions as gu) < 2")
```
Comment (C-Alexander) [inline app/dal/entities/Game.java]: is er een reden waarom je m gu noemt?

File: app/dal/entities/Game.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+})
+public class Game {
+    @OneToMany(cascade=CascadeType.ALL)
```
Comment (C-Alexander) [inline app/dal/entities/Game.java]: If I am not mistaken, this means that if a game ends, everyone in said game is logged out with their session deleted. 

Uhhh....

File: app/dal/entities/Game.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    @JoinTable(name = "Game_Sessions",
+            joinColumns = @JoinColumn(name = "gameId", referencedColumnName = "id"),
+            inverseJoinColumns = @JoinColumn(name = "sessionId", referencedColumnName = "id"))
+    private List<Session> sessions;
```
Comment (C-Alexander) [inline app/dal/entities/Game.java]: Can't hibernate just figure this out with @ManyToMany ?

File: app/dal/entities/Session.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package dal.entities;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name = "Session")
```
Comment (C-Alexander) [inline app/dal/entities/Session.java]: Please don't hardcode table names unless absolutely needed.

File: app/dal/entities/Session.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+public class Session {
+    @OneToOne(cascade=CascadeType.ALL)
```
Comment (C-Alexander) [inline app/dal/entities/Session.java]: Consider using the removeOrphans one too, it may be useful :)

File: app/dal/entities/Session.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinTable(name = "Game_Sessions",
```
Comment (C-Alexander) [inline app/dal/entities/Session.java]: Can't you handle this with a manytomany?

File: app/dal/entities/Session.java [hunk 23]
```diff
@@ -0,0 +1,34 @@
+package dal.entities;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name = "Session")
+public class Session {
```
Comment (C-Alexander) [inline app/dal/entities/Session.java]: Please rename this class, or it'll be super confusing with our session system. Or rename the session system I guess. Also consider if you really NEED sessions. Can't you just many-to-many games to Sessions?

The session-system sessions are unique for the current session the player is using (current device, a player can have say, 3 at once, theoretically), can be tracked to the player (session.getPlayer()) =)

Comment (C-Alexander) [reply app/dal/entities/Session.java]: also that is by far the fastest entity in our application - uses caching and multithreading!

File: conf/routes [hunk 24]
```diff
@@ -11,7 +11,7 @@ POST        /games               controllers.GameController.newGame
 GET         /game                controllers.GameController.game
 GET         /users               controllers.UserController.getUsers
 POST        /register            controllers.UserController.createUser
-POST        /login               controllers.UserController.login
+POST        /lobby               controllers.LobbyController.addToQueue
```
Comment (C-Alexander) [inline conf/routes]: why delete the login route?

**Comment by C-Alexander** [comment]
Decided to take a look as well wondering why it hadnt be accepted yet, found some more problems. Accepting it as is would break the application.

**Pull request merged**